### ISSUE TITLE
Rule proposal: disallow default values for fields that are never undefined

### ISSUE THREAD

**Issue by donaldpipowitch** [typescript-eslint/typescript-eslint#11405]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [x] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [x] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

During refactorings I sometimes encounter orphaned default values. This is dead code that creates noise and for complex type definitions they are hard to identify. I'm not aware of any rule or TypeScript setting that handle this.

### Fail Cases

```TypeScript
function example({ hello = 'world'}: { hello: string }) {}
```

### Pass Cases

```TypeScript
function example({ hello }: { hello: string }) {}
```

### Additional Info

I'd be happy to contribute this rule if I'm allowed to. ❤️

**Comment by Josh-Cena** [comment]
Yeah this is supposed to be caught by `no-unnecessary-condition` I'm surprised it doesn't

**Comment by JoshuaKGoldberg** [comment]
Agreed. This feature is so clearly good, I think it's actually a bug. 🙂

**Comment by kirkwaiblinger** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/6106 ?

**Comment by Josh-Cena** [comment]
Mmm indeed. I thought I searched
